nifi-api/src/main/java/org/apache/nifi/authorization/AccessPolicy.java

```diff
@@ -0,0 +1,125 @@
+package org.apache.nifi.authorization;
+
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Defines a policy for a set of entities to perform a set of actions on a given resource.
+ */
+public class AccessPolicy {
+
+    private final String identifier;
+
+    private final Resource resource;
+
+    private final Set<String> entities;
+
+    private final Set<RequestAction> actions;
+
+    /**
+     * Constructs a new policy with the given resource, entities, and actions.
+     *
+     * @param identifier the identifier of the policy
+     * @param resource the resource for the policy
+     * @param entities the entity ids for the policy (i.e. user or group ids)
+     * @param actions the actions for the policy
+     */
+    public AccessPolicy(final String identifier, final Resource resource, final Set<String> entities, final Set<RequestAction> actions) {
+        if (identifier == null || identifier.trim().isEmpty()) {
+            throw new IllegalArgumentException("Identifier can not be null or empty");
+        }
+
+        if (resource == null) {
+            throw new IllegalArgumentException("Resource can not be null");
+        }
+
+        if (entities == null || entities.isEmpty()) {
+            throw new IllegalArgumentException("Entities can not be null or empty");
+        }
+
+        if (actions == null || actions.isEmpty()) {
+            throw new IllegalArgumentException("Actions can not be null or empty");
+        }
+
+        this.identifier = identifier;
+        this.resource = resource;
+        this.entities = Collections.unmodifiableSet(entities);
+        this.actions = Collections.unmodifiableSet(actions);
+    }
+
+    /**
+     * @return the identifier for this policy
+     */
+    public String getIdentifier() {
+        return identifier;
+    }
+
+    /**
+     * @return the resource for this policy
+     */
+    public Resource getResource() {
+        return resource;
+    }
+
+    /**
+     * @return the set of entity ids for this policy
+     */
+    public Set<String> getEntities() {
+        return entities;
+    }
+
+    /**
+     * @return the set of actions for this policy
+     */
+    public Set<RequestAction> getActions() {
+        return actions;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+
+        final AccessPolicy other = (AccessPolicy) obj;
+
+        return this.identifier.equals(other.getIdentifier())
+                && this.resource.getIdentifier().equals(other.getResource().getIdentifier())
+                && this.entities.equals(other.entities)
+                && this.actions.equals(other.actions);
+    }
+
+    @Override
+    public int hashCode() {
+        int hash = 7;
+        hash = 53 * hash + Objects.hash(this.identifier, this.resource, this.entities, this.actions);
```

Comment:
When using Objects.hash(), you should be able to just return the invocation of that with the class members you want included in the hash:
`return Objects.hash(this.identifier, this.resource, this.entities, this.actions);`